### ISSUE TITLE
Parallel Processing

### ISSUE THREAD

**Issue by itdependsnetworks** [networktocode/diffsync#202]
### Environment
* DiffSync version:  latest

### Proposed Functionality

The ability to process parts of the sync that do not effect each other.

This likely requires a facility that developers can describe what can be parallel as well as potentially finding specific scenarios where it can automagically happen. 

### Use Case

Use cases that would benefit from this include:

* Getting Side A and Side B at the same time
* Getting all of table 1 and all of table 2 from Side A at the same time
* Update the model where there is no relationships at the same time
* A parent of multiple children, each of those children can run at the same time

**Comment by chadell** [comment]
isn't this something you could infer based on the children relationships? I mean, all the object of one type, e.g `site`, should be able to be processed in parallel. Obviously, you can find shared children, but this conflict can be postponed for a final merge.
Don't know if assuming, by default, that the objects of parent models can run in parallel makes sense. Then, if the library offers a remediation final step using the already supported Redis memory, you could run multiprocessing.
does it make sense to you? or maybe missing some point?

**Comment by itdependsnetworks** [comment]
Some can be inferred, some may. It be able to be inferred, but yes, that is the point

**Comment by itdependsnetworks** [comment]
If I am syncing all of my locations and all of my devices, I should be able to run those both at the same time, as an example.

**Comment by jdrew82** [comment]
> If I am syncing all of my locations and all of my devices, I should be able to run those both at the same time, as an example.

But that's nothing to do with DiffSync itself. DiffSync is simply comparing the data once you've loaded it in. How you load it in is up to you and that can easily be done in parallel. In fact, I recently updated SSoT to do just this.

**Comment by itdependsnetworks** [comment]
Yea, that is basically it. I don't remember the originating reason this came up tbh. Overall, I think it is a good improvement, but likely easier areas to address with larger impact, before this one.
